dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1183 @@
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 1,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Header: &types.Header{
+					Number: 2,
+				},
+				Body: &types.Body{},
+			},
+		},
+	}
+
+	hash := (&types.Header{
+		Number: 2,
+	}).Hash()
+	err = cs.validateResponse(req, resp, "")
+	require.Equal(t, errResponseIsNotChain, err)
+	require.True(t, cs.pendingBlocks.hasBlock(hash))
+	cs.pendingBlocks.removeBlock(hash)
+
+	parent := (&types.Header{
+		Number: 1,
+	}).Hash()
+	header3 := &types.Header{
+		ParentHash: parent,
+		Number:     3,
+	}
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 1,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Hash:          header3.Hash(),
+				Header:        header3,
+				Body:          &types.Body{},
+				Justification: &[]byte{0},
+			},
+		},
+	}
+
+	hash = (&types.Header{
+		ParentHash: parent,
+		Number:     3,
+	}).Hash()
+	err = cs.validateResponse(req, resp, "")
+	require.Equal(t, errResponseIsNotChain, err)
+	require.True(t, cs.pendingBlocks.hasBlock(hash))
+	bd := cs.pendingBlocks.getBlock(hash)
+	require.NotNil(t, bd.justification)
+	cs.pendingBlocks.removeBlock(hash)
+
+	parent = (&types.Header{
+		Number: 2,
+	}).Hash()
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 2,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Header: &types.Header{
+					ParentHash: parent,
+					Number:     3,
+				},
+				Body: &types.Body{},
+			},
+		},
+	}
+
+	err = cs.validateResponse(req, resp, "")
+	require.NoError(t, err)
+	require.False(t, cs.pendingBlocks.hasBlock(hash))
+
+	req = &network.BlockRequestMessage{
+		RequestedData: network.RequestedDataJustification,
+	}
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Justification: &[]byte{0},
+			},
+		},
+	}
+
+	err = cs.validateResponse(req, resp, "")
+	require.NoError(t, err)
+	require.False(t, cs.pendingBlocks.hasBlock(hash))
+}
+
+func TestChainSync_validateResponse_firstBlock(t *testing.T) {
+	t.Parallel()
+
+	cs := newTestChainSync(t)
+	bs := new(mocks.BlockState)
+	bs.On("HasHeader", mock.AnythingOfType("common.Hash")).Return(false, nil)
+	cs.blockState = bs
+
+	req := &network.BlockRequestMessage{
+		RequestedData: bootstrapRequestData,
+	}
+
+	header := &types.Header{
+		Number: 2,
+	}
+
+	resp := &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Hash: header.Hash(),
+				Header: &types.Header{
+					Number: 2,
+				},
+				Body:          &types.Body{},
+				Justification: &[]byte{0},
+			},
+		},
+	}
+
+	err := cs.validateResponse(req, resp, "")
+	require.True(t, errors.Is(err, errUnknownParent))
+	require.True(t, cs.pendingBlocks.hasBlock(header.Hash()))
+	bd := cs.pendingBlocks.getBlock(header.Hash())
+	require.NotNil(t, bd.header)
+	require.NotNil(t, bd.body)
+	require.NotNil(t, bd.justification)
```

Comment:
Can you assert the full `bd` value instead of just checking for non nil for some fields? 🤔

Reply:
Updated.